File: year3/sem1/ppd/lab/lab4/secv.cpp
#include <iostream>
#include <fstream>

using namespace std;

int** create_mat(const int& n, const int& m) {
    int** mat = new int*[n + 1];
    for(int i = 0 ;i <= n;++i) {
        mat[i] = new int[m + 1];
    }
    return mat;
}

void delete_mat(const int& n, const int& m, int** mat) {
    for(int i = 0;i < n;++i) {
        delete[] mat[i];
    }
    delete[] mat;
}

void read_mat(const int& n, const int& m, int** mat, ifstream& f) {
    for(int i = 1;i <= n;++i) {
        for(int j = 1;j <= m;++j) {
            f>>mat[i][j];
        }
    }
}

int conv(const int& i, const int& j, const int& n_mat, const int& m_mat, const int& n_conv, const int& m_conv, int** mat, int** conv_mat) {
    int sum = 0;
    for(int row = i - (n_conv / 2);row <= i + (n_conv / 2);++row) {
        for(int column = j - (m_conv / 2);column <= j + (m_conv / 2);++column) {
            sum += mat[min(n_mat, max(1, row))][min(m_mat, max(1, column))] * conv_mat[row - i + (n_conv / 2) + 1][column - j + (m_conv / 2) + 1];
        }
    }

    return sum;
}

void apply_conv(const int& n_mat, const int& m_mat, const int& n_conv, const int& m_conv, int** mat, int** conv_mat, int** result) {
    for(int i = 1;i <= n_mat;++i) {
        for(int j = 1;j <= m_mat;++j) {
            result[i][j] = conv(i, j, n_mat, m_mat, n_conv, m_conv, mat, conv_mat);
            cout<<result[i][j]<<'\n';
        }
    }
}

void print_mat(const int& n, const int& m, int** mat, const char* filename) {
    ofstream output(filename);

    output<<n<<" "<<m<<'\n';
    for(int i = 1;i <= n;++i, output<<'\n') {
        for(int j = 1;j <= m;++j) {
            output<<mat[i][j]<<" ";
        }
    }
    output.close() ;
}

int main(int argc, char* argv[]) {
    if(argc != 4) {
        cout<<"Incorect call: <input_file> <conv_mat_file> <output_file>\n";
        return 0;
    }

    ifstream input_mat(argv[1]);
    ifstream input_conv(argv[2]);

    int n_mat, m_mat, n_conv, m_conv;
    input_mat>>n_mat>>m_mat;
    input_conv>>n_conv>>m_conv;
    int** mat = create_mat(n_mat + 1, m_mat + 1);
    int** conv_mat = create_mat(n_conv + 1, m_conv + 1);
    int** result_mat = create_mat(n_mat + 1, m_mat + 1);

    read_mat(n_mat, m_mat, mat, input_mat);
    read_mat(n_conv, m_conv, conv_mat, input_conv);
    input_mat.close();
    input_conv.close();

    apply_conv(n_mat, m_mat, n_conv, m_conv, mat, conv_mat, result_mat);
    print_mat(n_mat, m_mat, result_mat, argv[3]);

    delete_mat(n_mat + 1, m_mat + 1, mat);
    delete_mat(n_conv + 1, m_conv + 1, conv_mat);
    delete_mat(n_mat + 1, m_mat + 1, result_mat);

    return 0;
}

File: year3/sem1/ppd/lab/lab4/Secv.java
import java.io.File;
import java.io.FileWriter;
import java.util.Scanner;

public class Secv {
    public static void main(String[] args) {
        if(args.length != 3) {
            System.out.println("Usage: java Main <input_file> <conv_mat_file> <output_file>");
            System.exit(1);
        }

        String matFilePath = args[0];
        String convFilePath = args[1];
        String outputFilePath = args[2];

        try(Scanner scanner = new Scanner(new File(matFilePath))) {
            nMat = scanner.nextInt();
            mMat = scanner.nextInt();
            mat = new int[nMat][mMat];
            result_mat = new int[nMat][mMat];
            
            for(int i = 0;i < nMat;++i) {
                for(int j = 0 ;j < mMat;++j) {
                    mat[i][j] = scanner.nextInt();
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
        }

        try(Scanner scanner = new Scanner(new File(convFilePath))) {
            nConv = scanner.nextInt();
            mConv= scanner.nextInt();
            conv_mat = new int[nConv][mConv];
            
            for(int i = 0;i < nConv;++i) {
                for(int j = 0 ;j < mConv;++j) {
                    conv_mat[i][j] = scanner.nextInt();
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        applyConv();

        try(FileWriter writer = new FileWriter(outputFilePath)) {
            writer.write(nMat + " " + mMat + "\n");
            for(int i = 0;i < nMat;++i, writer.write("\n")) {
                for(int j = 0;j < mMat;++j) {
                    writer.write(result_mat[i][j] + " ");
                }
            }
        } catch(Exception e) {
            e.printStackTrace();
        }
    }

    private static int conv(int i, int j) {
        int sum = 0;
        for(int row = i - (nConv / 2);row <= i + (nConv / 2);++row) {
            for(int column = j - (mConv / 2);column <= j + (mConv / 2);++column) {
                sum += mat[Math.min(nMat - 1, Math.max(0, row))][Math.min(mMat - 1, Math.max(0, column))] * conv_mat[row - i + (nConv / 2)][column - j + (mConv / 2)];
            }
        }
        return sum;
    }

    private static void applyConv() {
        for(int i = 0;i < nMat;++i) {
            for(int j = 0;j < mMat;++j) {
                result_mat[i][j] = conv(i, j);
            }
        }
    }

    private static int nMat, mMat, nConv, mConv;
    private static int[][] mat, conv_mat, result_mat; 
}